runner/src/runner.rs

```diff
@@ -0,0 +1,631 @@
+    // Ensure there is no other child running
+    if runner.child_proc().is_some() {
+        return Err(Error::ChildProcessExists);
+    }
+    let downloaded_release = runner.downloaded_release().as_ref().ok_or(Error::NoDownloadedRelease)?;
+    let child = Command::new(downloaded_release.path.as_os_str())
+        .args(runner.vault_args().clone())
+        .stdout(stdout_mode)
+        .spawn()?;
+    Ok(child)
+}
+
+#[cfg(test)]
+mod tests {
+    use async_trait::async_trait;
+    use bytes::Bytes;
+    use codec::Decode;
+
+    use sp_core::{Bytes as SpCoreBytes, H256};
+
+    use std::{
+        convert::TryInto,
+        fmt::Debug,
+        fs::{self, File},
+        io::Write,
+        os::unix::prelude::PermissionsExt,
+        path::PathBuf,
+        process::{Child, Command, Stdio},
+        str::FromStr,
+    };
+
+    use crate::error::Error;
+
+    use super::*;
+
+    use sysinfo::{Pid, System, SystemExt};
+
+    mockall::mock! {
+        Runner {}
+
+        #[async_trait]
+        pub trait RunnerExt {
+            fn parachain_rpc(&self) -> &WebsocketClient;
+            fn vault_args(&self) -> &Vec<String>;
+            fn child_proc(&mut self) -> &mut Option<Child>;
+            fn set_child_proc(&mut self, child_proc: Child);
+            fn downloaded_release(&self) -> &Option<DownloadedRelease>;
+            fn set_downloaded_release(&mut self, downloaded_release: DownloadedRelease);
+            fn download_path(&self) -> &PathBuf;
+            fn set_download_path(&mut self, download_path: PathBuf);
+            async fn try_get_release(&self, pending: bool) -> Result<Option<ClientRelease>, Error>;
+            async fn download_binary(&mut self, release: ClientRelease) -> Result<(), Error>;
+            fn uri_to_bin_path(&self, uri: &str) -> Result<(String, PathBuf), Error>;
+            fn delete_downloaded_release(&mut self) -> Result<(), Error>;
+            async fn run_binary(&mut self) -> Result<(), Error>;
+            fn terminate_proc_and_wait(&mut self) -> Result<(), Error>;
+            async fn get_request_bytes(&self, url: String) -> Result<Bytes, Error>;
+        }
+
+        #[async_trait]
+        pub trait StorageReader {
+            fn compute_storage_key(&self, module: String, key: String) -> String;
+            async fn query_storage(
+                &self,
+                parachain_rpc: &WebsocketClient,
+                maybe_storage_key: Option<String>,
+                method: String,
+            ) -> Option<SpCoreBytes>;
+            async fn read_chain_storage<T: 'static + Decode + Debug>(
+                &self,
+                parachain_rpc: &WebsocketClient,
+                maybe_storage_key: Option<String>,
+            ) -> Result<Option<T>, Error>;
+        }
+    }
+
+    #[tokio::test]
+    async fn test_runner_download_binary() {
+        let mut runner = MockRunner::default();
+        let mock_path = PathBuf::from_str("./vault-standalone-metadata").unwrap();
```

Comment:
We should probably use `TempDir` for all of these files (see `runtime/src/integration/mod.rs`) - if the test gets terminated the file is not cleaned up, leading to clutter and potentially failing tests